src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs

```diff
@@ -144,12 +144,12 @@ public bool MoveNext()
 
                         if (isDirectory && !isSpecialDirectory)
                         {
-                            if (_options.RecurseSubdirectories && ShouldRecurseIntoEntry(ref entry))
+                            if (_options.RecurseSubdirectories && _remainingDepth > 0 && ShouldRecurseIntoEntry(ref entry))
```

Comment:
I know from experience that any changes in `FileSystemEnumerator.MoveNext` could cause ugly performance regressions.

Can you please run the enumeration performance benchmarks, before and after your changes, to see if there's an impact?

In case you've not run benchmarks before, here are the nstructions: [dotnet/performance/docs/benchmarking-workflow-dotnet-runtime.md#preventing-regressions](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md#preventing-regressions)

These are the relevant benchmarks: [dotnet/performance/src/benchmarks/micro/libraries/System.IO.FileSystem/Perf.Directory.cs](https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/libraries/System.IO.FileSystem/Perf.Directory.cs).

I am particularly interested in the results for `RecursiveDepthData()` and `EnumerateFiles()`.

Reply:
I will try to run these tests, but it may take a long time since I have never done this.
Perhaps @adamsitnik  will get ahead of me. :-)

Reply:
After my first attempt no regression was found. I am not sure I did everything is correct.

Reply:
Can you please share the results here?

Reply:
I will do but I am not still sure I make right steps to make the work. Docs are not so clear about iterations. If an expert like @adamsitnik published a short (5 minutes) video with demo steps how clean start, make a base and iterate with changes it would be great help (because each iteration takes some hours on my notebook).

Currently I can share only "No differences found between the benchmark results with threshold 1%.". Do you want to get an archive with result files?

Reply:
I attached the archive with the results after doing it more neatly
[1.zip](https://github.com/dotnet/runtime/files/6004732/1.zip)

Reply:
There is very good documentation in the dotnet/performance repo. You should end up with a file containing markdown tables that you can paste directly into your PR here.

Reply:
Also it should take no more than a few minutes to run the relevant benchmarks. You can just filter to the file enumeration ones.

Reply:
I extracted your zip file and compared the results using ResultComparer from the performance repo:

```
D:\performance\src\tools\ResultsComparer> dotnet run --base "C:\Users\carlos\Desktop\New folder\before" --diff "C:\Users\carlos\Desktop\New folder\after" --threshold 0.00001%
summary:
better: 1, geomean: 1.016
total diff: 1

No Slower results for the provided threshold = 0.00001% and noise filter = 0.3ns.

| Faster                                        | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| --------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.IO.Tests.Perf_Directory.EnumerateFiles |      1.02 |       8712455.56 |       8576044.64 |         |
```

@iSazonov can you run a benchmark test for the other enumeration-related method `RecursiveDepthData()`? Or even better: just run all the methods in that perf class.

Reply:
@carlossanlop I don't see regressions but I can not trust myself and you can not too because I never run perf tests before (I am not sure I do all right). In any case we need a confirmation from more experienced developer.

Reply:
All results for System.IO.Tests.Perf_Directory.*
[2.zip](https://github.com/dotnet/runtime/files/6091956/2.zip)